server/routes/induction/update/personalInterestsUpdateController.test.ts

```diff
@@ -358,5 +358,48 @@ describe('personalInterestsUpdateController', () => {
       expect(req.session.personalInterestsForm).toEqual(personalInterestsForm)
       expect(req.session.inductionDto).toEqual(inductionDto)
     })
+
+    it('should update induction DTO and redirect back to check your answers page when coming from check your answers', async () => {
+      // Given
+      const prisonerSummary = aValidPrisonerSummary()
+      req.session.prisonerSummary = prisonerSummary
+      req.session.inductionDto = aLongQuestionSetInductionDto()
+      req.session.pageFlowHistory = {
+        pageUrls: [
+          `/prisoners/${prisonerSummary.prisonNumber}/induction/check-your-answers`,
+          `/prisoners/${prisonerSummary.prisonNumber}/induction/personal-interests`,
+        ],
+        currentPageIndex: 1,
+      }
+
+      req.body = {
+        personalInterests: ['KNOWLEDGE_BASED', 'OTHER'],
+        personalInterestsOther: 'Writing poetry and short stories',
+      }
+
+      // The actual implementations are fine for this test
```

Comment:
Assuming I've read this correctly .... I'm not sure about this .... 🤔 

* We are testing a class (`PersonalInterestsUpdateController` in this case)
* Because we want our test isolated to the class/method, we mock the class's dependencies (in this case the mapper and the form validator)
* Then we take the mock, and replace the method/function that the code under test (controller method) will call with the real implementation 🤔 

I don't know how well this sits with me. Coming from a java/kotlin background, the frameworks in question (mockito) don't really allow you to do this (the closest you have is a Spy, but thats used for a different reason).
Maybe it's because of that background. Perhaps this type of approach is more accepted in js/typescript worlds? 🤷‍♂️ 

I'm happy to go with this as it is; I'm no expert in these matters 😁  and am all up for learning new or different ways of doing things 👍 

But I think if what we are saying here is that we are happy with the default implementation of a given function, rather than replace it inline in the mock like this, would we be better off simply not using a mock at all for the dependency, and simply using the real thing?

Something to think about moving forwards 🤔 👍

Reply:
Yeah, it might be a background thing. Personally I'm not keen on mocking unless there's a reason to do so and I couldn't see any reason to use a mock here. The more mocks you use the more tests are testing the implementations of mocks rather than the code, plus it's extra code to maintain mock implementations. So for that reason I prefer limiting mocks and testing against live code where possible.

The behaviour of these mapper and validator helper methods should be predictable, and if they ever did do something the controller method doesn't expect then using a mock would only reduce coverage in catching those bugs.

Additionally, there is no alternative implementation here. These mapper and validator helper methods are always explicitly imported and implemented in the controller logic. These aren't injected as dependancies where different mappers/validators could be used depending on the implementation of the controller.

That was my thinking anyway. 🤷‍♂️